FAQ: What types does my pullback need to accept? (#428)
JuliaDiff/ChainRulesCore.jl: docs/src/writing_good_rules.md

## Ensure your pullback can accept the right types
As a rule the number of types you need to accept in a pullback is theoretically unlimitted, but practically highly constrained to inline with the primal return type.
The three kinds of inputs you will practically need to accept one or more of: _natural tangents_, _structural tangents_, and _thunks_.
You do not in general have to handle `AbstractZero`s as the AD system will not call the pullback if the input is a zero, since the output will also be.
Some more background information on these types can be found in [the design notes](@ref manytypes).
In many cases these all all tangents can be treated the same: tangent types overload a bunch of linear-operators, and the majority of functions used inside a pullback are linear operators.
If you find linear operators from Base/stdlibs that are not supported, consider openning an issue or PR on the [ChainRulesCorejl repo](https://github.com/JuliaDiff/ChainRulesCore.jl/).

### Natural tangents
Natural tangent types are the types you might feel the tangent should be.
These are a purely human notion, they are the types the user wants to use because they make the math easy.

> If a primal type `P` overloads subtraction (`-(::P,::P)`) then that generally returns a  natural tangent type for `P`; but this is not required to be defined.

Reviewer: ```suggestion
There is currently no formal definition of what consistutes a natural tangent, but there are a few heuristics.
For example, if a primal type `P` overloads subtraction (`-(::P,::P)`) then that generally returns a  natural tangent type for `P`; but this is not required to be defined.
```

For example, for most `AbstractArray` subtypes, most other `AbstractArray`s (of right size and element type) can be considered as natural tangent types.

Not all types have natural tangent types.
For example there is no natural differential for a `Tuple`.
It is not a `Tuple` since that doesn't have any method for `+`.
Similar is true for many `struct`s.
For those cases there is only a structural differential.

### Structural tangents

Structural tangents are tangent types that shadow the structure of the primal type.
They are represented by the [`Tangent`](@ref) type.
They can represent any composite type, such as a tuple, or a structure (or a `NamedTuple`) etc.

!!! info "Do I have to support the structural tangents as well?"
    Technically, you might not actually have to write rules to accept structural tangents; if the AD system never has to decompose down to the level of getfield.
    This is common for types that don't support user getfield/getproperty access, and that have a lot of rules for the ways they are accessed (such cases include many `AbstractArray` subtypes).
    You likely should support it just in case.

>     But if it is causing struggles, then you can leave it off til someone complains.

Reviewer: I would really rather we didn't say this, per discussion on slack etc.

The main argument is that any time AD runs successfully on code that accesses the fields of an AbstractArray, a structural tangent will be obtained.

Author: What if changed it from 

> You likely should support it just in case.

to

> You **really** should support it just in case; especially if the primal type in question is not restricted to a well-tested concrete type.

Reviewer: I still don't think this captures it. I would prefer something like the following:

You really should support it, because it's what AD will naturally produce if code involving `getfield` is ever differentiated through.

Author: I think that is covered by the previous sentence.

```
to define the rules.

## Ensure your pullback can accept the right types

> As a rule the number of types you need to accept in a pullback is theoretically unlimitted, but practically highly constrained to inline with the primal return type.

Reviewer 2: ```suggestion
As a rule the number of types you need to accept in a pullback is theoretically unlimitted, but practically highly constrained to be in line with the primal return type.
```

The three kinds of inputs you will practically need to accept one or more of: _natural tangents_, _structural tangents_, and _thunks_.
You do not in general have to handle `AbstractZero`s as the AD system will not call the pullback if the input is a zero, since the output will also be.
Some more background information on these types can be found in [the design notes](@ref manytypes).

> In many cases these all all tangents can be treated the same: tangent types overload a bunch of linear-operators, and the majority of functions used inside a pullback are linear operators.

Reviewer 2: ```suggestion
In many cases all these tangents can be treated the same: tangent types overload a bunch of linear-operators, and the majority of functions used inside a pullback are linear operators.
```

> If you find linear operators from Base/stdlibs that are not supported, consider openning an issue or PR on the [ChainRulesCorejl repo](https://github.com/JuliaDiff/ChainRulesCore.jl/).

Reviewer 2: ```suggestion
If you find linear operators from Base/stdlibs that are not supported, consider opening an issue or a PR on the [ChainRulesCore.jl repo](https://github.com/JuliaDiff/ChainRulesCore.jl/).
```

> ### Natural tangents

Reviewer 2: Not objecting to adding this, but don't we already say what the differences between natural and structural tangents are somewhere in the docs?

Author: In the design docs we probably do.
I am in-favour of having redundancy in the docs.
It makes them much more understandable, since noone reads them end-to-end.
Albeit, at the cost of making the docs harder to maintain.

> Natural tangent types are the types you might feel the tangent should be.

Author: ```suggestion
Natural tangent types are the types you might feel the tangent should be, to represent a small change in the primal value.
For example, if the primal is a `Float32`, the natural tangent is also a `Float32`.
Slightly more complex, for a `ComplexF64` the natural tangent is again also a `ComplexF64`, we almost never want to use the structural tangent `Tangent{ComplexF64}(re=..., im=...)` which is defined.
For other cases, this gets a little more complicated, see below.
```

> These are a purely human notion, they are the types the user wants to use because they make the math easy.

Reviewer 3: I think this is a bit vague. Even if you don't provide an all-encompassing definition, you need to tell the reader roughly what you mean by "natural" at all, what distinction are you drawing? If they don't already know then all these disclaimers are pretty confusing.

I would do this by example. The simplest one would be `ComplexF64`, that's a struct, and by "natural" you mean a gradient which is another number, as opposed to "structural" you mean `Tangent{Complex}(...)`. And after that, move on to more complicated ones.

Author: Are there not several examples in the next section?

Reviewer 3: There are, but you don't quite spell out what this word "natural" is being used to mean, before using it a bunch of times. I think that should go first: first define the terminology, then the clear-cut cases on both sides, then the tricky cases.

Author: ok fair.

> There is currently no formal definition of what consistutes a natural tangent, but there are a few heuristics.

Reviewer 2: ```suggestion
There is currently no formal definition of what constitutes a natural tangent, but there are a few heuristics.
```

> For example, if a primal type `P` overloads subtraction (`-(::P,::P)`) then that generally returns a  natural tangent type for `P`; but this is not required to be defined.

Reviewer 2: ```suggestion
For example, if a primal type `P` overloads subtraction (`-(::P,::P)`) then that generally returns a natural tangent type for `P`; but this is not required to be defined.
```

> For example, if a primal type `P` overloads subtraction (`-(::P,::P)`) then that generally returns a natural tangent type for `P`; but this is not required to be defined.

Author: ```suggestion
For example, if a primal type `P` overloads subtraction (`-(::P,::P)`) then that generally returns a natural tangent type for `P`; but this is not required to be defined and sometimes it is defined poorly.
```

Another example is for a `DateTime`, any `Period` subtype, such as `Millisecond` or `Nanosecond` is a natural differential.
There are often many different natural tangent types for a given primal type.
However, they are generally closely related and duck-type the same.
For example, for most `AbstractArray` subtypes, most other `AbstractArray`s (of right size and element type) can be considered as natural tangent types.

Not all types have natural tangent types.
For example there is no natural differential for a `Tuple`.
It is not a `Tuple` since that doesn't have any method for `+`.
Similar is true for many `struct`s.
For those cases there is only a structural differential.

### Structural tangents

Structural tangents are tangent types that shadow the structure of the primal type.
They are represented by the [`Tangent`](@ref) type.
They can represent any composite type, such as a tuple, or a structure (or a `NamedTuple`) etc.

!!! info "Do I have to support the structural tangents as well?"

>     Technically, you might not actually have to write rules to accept structural tangents; if the AD system never has to decompose down to the level of getfield.

Reviewer 2: ```suggestion
    Technically, you might not actually have to write rules to accept structural tangents; if the AD system never has to decompose down to the level of `getfield`.
```

>     This is common for types that don't support user getfield/getproperty access, and that have a lot of rules for the ways they are accessed (such cases include some `AbstractArray` subtypes).

Reviewer 2: ```suggestion
    This is common for types that don't support user `getfield`/`getproperty` access, and that have a lot of rules for the ways they are accessed (such cases include some `AbstractArray` subtypes).
```

>     This is common for types that don't support user getfield/getproperty access, and that have a lot of rules for the ways they are accessed (such cases include many `AbstractArray` subtypes).
>     You likely should support it just in case.

Author: ```suggestion
    This is common for types that don't support user getfield/getproperty access, and that have a lot of rules for the ways they are accessed (such cases include some `AbstractArray` subtypes).
    You really should support it just in case; especially if the primal type in question is not restricted to a well-tested concrete type.
```

    But if it is causing struggles, then you can leave it off til someone complains.

### Thunks

A thunk (either a [`Thunk`](@ref), or a [`InplaceableThunk`](@ref)), represents a delayed computation.
They can be thought of as a wrapper of the value the computation returns.
In this sense they wrap either a natural or structural tangent.

> !!! warning "You should to support AbstractThunk inputs even if you don't use thunks"

Reviewer 2: ```suggestion
!!! warning "You should support `AbstractThunk` inputs even if you don't use thunks"
```

Author: This headings of admonitions don't support markdown, as I recall

>      Unfortunately the AD sytems do not know what rules support thunks and what do not.

Reviewer 2: ```suggestion
     Unfortunately the AD sytems do not know which rules support thunks and which do not.
```
